{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "15885164",
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(\"D:/Bioinfomation_Project/Create_R_Package\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e4b1d52b",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(warn = -1)\n",
    "suppressMessages(library(tidyverse))\n",
    "suppressMessages(library(CARD))\n",
    "suppressMessages(library(data.table))\n",
    "suppressMessages(library(Seurat))\n",
    "suppressMessages(library(spacexr))\n",
    "suppressMessages(library(Giotto))\n",
    "suppressMessages(library(future))\n",
    "suppressMessages(library(dplyr))\n",
    "suppressMessages(library(SPOTlight))\n",
    "suppressMessages(library(igraph))\n",
    "suppressMessages(library(SeuratDisk))\n",
    "suppressMessages(library(patchwork))\n",
    "suppressMessages(library(Matrix))\n",
    "suppressMessages(library(data.table))\n",
    "suppressMessages(library(spdep))\n",
    "suppressMessages(library(spatialreg))\n",
    "suppressMessages(library(gridExtra))\n",
    "suppressMessages(library(msigdbr))\n",
    "suppressMessages(library(fields))\n",
    "suppressMessages(library(Spatialsmooth))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e4de84b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running CARD ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "## QC on scRNASeq dataset! ...\n",
      "## QC on spatially-resolved dataset! ...\n",
      "## create reference matrix from scRNASeq...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: nnls\n",
      "\n",
      "Loading required package: TOAST\n",
      "\n",
      "Loading required package: EpiDISH\n",
      "\n",
      "Loading required package: limma\n",
      "\n",
      "\n",
      "Attaching package: 'limma'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:BiocGenerics':\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "Loading required package: quadprog\n",
      "\n",
      "Registered S3 method overwritten by 'GGally':\n",
      "  method from   \n",
      "  +.gg   ggplot2\n",
      "\n",
      "MuSiC v1.0.0 support SingleCellExperiment! See Tutorial: https://xuranw.github.io/MuSiC/articles/MuSiC.html\n",
      "\n",
      "MuSiC2 for multi-condition bulk RNA-seq data is also available!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "## Select Informative Genes! ...\n",
      "## Deconvolution Starts! ...\n",
      "## Deconvolution Finish! ...\n",
      "CARD completed successfully.\n",
      "Running RCTD ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"prepare data\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Begin: process_cell_type_info\n",
      "\n",
      "process_cell_type_info: number of cells in reference: 1926\n",
      "\n",
      "process_cell_type_info: number of genes in reference: 19736\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "                        Acinar_cells                       Cancer_clone_A \n",
      "                                  13                                  126 \n",
      "                      Cancer_clone_B            Ductal_APOL1_high-hypoxic \n",
      "                                 170                                  215 \n",
      "Ductal_CRISP3_high-centroacinar_like                  Ductal_MHC_Class_II \n",
      "                                 529                                  287 \n",
      "         Ductal_terminal_ductal_like                      Endocrine_cells \n",
      "                                 350                                    3 \n",
      "                   Endothelial_cells                          Fibroblasts \n",
      "                                  11                                    5 \n",
      "                       Macrophages_A                        Macrophages_B \n",
      "                                  21                                   19 \n",
      "                          Mast_cells                               mDCs_A \n",
      "                                  14                                   12 \n",
      "                              mDCs_B                            Monocytes \n",
      "                                  33                                   18 \n",
      "                                pDCs                                 RBCs \n",
      "                                  13                                   15 \n",
      "                  T_cells_&_NK_cells                           Tuft_cells \n",
      "                                  40                                   32 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "End: process_cell_type_info\n",
      "\n",
      "create.RCTD: getting regression differentially expressed genes: \n",
      "\n",
      "get_de_genes: Acinar_cells found DE genes: 57\n",
      "\n",
      "get_de_genes: Cancer_clone_A found DE genes: 324\n",
      "\n",
      "get_de_genes: Cancer_clone_B found DE genes: 342\n",
      "\n",
      "get_de_genes: Ductal_APOL1_high-hypoxic found DE genes: 330\n",
      "\n",
      "get_de_genes: Ductal_CRISP3_high-centroacinar_like found DE genes: 225\n",
      "\n",
      "get_de_genes: Ductal_MHC_Class_II found DE genes: 259\n",
      "\n",
      "get_de_genes: Ductal_terminal_ductal_like found DE genes: 264\n",
      "\n",
      "get_de_genes: Endocrine_cells found DE genes: 769\n",
      "\n",
      "get_de_genes: Endothelial_cells found DE genes: 474\n",
      "\n",
      "get_de_genes: Fibroblasts found DE genes: 709\n",
      "\n",
      "get_de_genes: Macrophages_A found DE genes: 331\n",
      "\n",
      "get_de_genes: Macrophages_B found DE genes: 390\n",
      "\n",
      "get_de_genes: Mast_cells found DE genes: 335\n",
      "\n",
      "get_de_genes: mDCs_A found DE genes: 386\n",
      "\n",
      "get_de_genes: mDCs_B found DE genes: 282\n",
      "\n",
      "get_de_genes: Monocytes found DE genes: 478\n",
      "\n",
      "get_de_genes: pDCs found DE genes: 556\n",
      "\n",
      "get_de_genes: RBCs found DE genes: 18\n",
      "\n",
      "get_de_genes: T_cells_&_NK_cells found DE genes: 390\n",
      "\n",
      "get_de_genes: Tuft_cells found DE genes: 406\n",
      "\n",
      "get_de_genes: total DE genes: 4030\n",
      "\n",
      "create.RCTD: getting platform effect normalization differentially expressed genes: \n",
      "\n",
      "get_de_genes: Acinar_cells found DE genes: 75\n",
      "\n",
      "get_de_genes: Cancer_clone_A found DE genes: 814\n",
      "\n",
      "get_de_genes: Cancer_clone_B found DE genes: 793\n",
      "\n",
      "get_de_genes: Ductal_APOL1_high-hypoxic found DE genes: 686\n",
      "\n",
      "get_de_genes: Ductal_CRISP3_high-centroacinar_like found DE genes: 554\n",
      "\n",
      "get_de_genes: Ductal_MHC_Class_II found DE genes: 618\n",
      "\n",
      "get_de_genes: Ductal_terminal_ductal_like found DE genes: 617\n",
      "\n",
      "get_de_genes: Endocrine_cells found DE genes: 1330\n",
      "\n",
      "get_de_genes: Endothelial_cells found DE genes: 1123\n",
      "\n",
      "get_de_genes: Fibroblasts found DE genes: 1281\n",
      "\n",
      "get_de_genes: Macrophages_A found DE genes: 671\n",
      "\n",
      "get_de_genes: Macrophages_B found DE genes: 812\n",
      "\n",
      "get_de_genes: Mast_cells found DE genes: 732\n",
      "\n",
      "get_de_genes: mDCs_A found DE genes: 881\n",
      "\n",
      "get_de_genes: mDCs_B found DE genes: 563\n",
      "\n",
      "get_de_genes: Monocytes found DE genes: 899\n",
      "\n",
      "get_de_genes: pDCs found DE genes: 1231\n",
      "\n",
      "get_de_genes: RBCs found DE genes: 39\n",
      "\n",
      "get_de_genes: T_cells_&_NK_cells found DE genes: 962\n",
      "\n",
      "get_de_genes: Tuft_cells found DE genes: 970\n",
      "\n",
      "get_de_genes: total DE genes: 6351\n",
      "\n",
      "fitBulk: decomposing bulk\n",
      "\n",
      "chooseSigma: using initial Q_mat with sigma =  1\n",
      "\n",
      "Likelihood value: 687228.018597623\n",
      "\n",
      "Sigma value:  0.84\n",
      "\n",
      "Likelihood value: 682625.148132922\n",
      "\n",
      "Sigma value:  0.69\n",
      "\n",
      "Likelihood value: 680550.270825191\n",
      "\n",
      "Sigma value:  0.66\n",
      "\n",
      "Likelihood value: 680435.391127476\n",
      "\n",
      "Sigma value:  0.65\n",
      "\n",
      "Likelihood value: 680420.862886743\n",
      "\n",
      "Sigma value:  0.64\n",
      "\n",
      "Likelihood value: 680418.540217078\n",
      "\n",
      "Sigma value:  0.64\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RCTD completed successfully.\n",
      "Running SPOTlight ...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Validating h5Seurat file\n",
      "\n",
      "Initializing RNA with data\n",
      "\n",
      "Adding counts for RNA\n",
      "\n",
      "Adding miscellaneous information for RNA\n",
      "\n",
      "Adding command information\n",
      "\n",
      "Adding cell-level metadata\n",
      "\n",
      "Adding miscellaneous information\n",
      "\n",
      "Adding tool-specific results\n",
      "\n",
      "Calculating cluster Acinar_cells\n",
      "\n",
      "Calculating cluster Ductal_terminal_ductal_like\n",
      "\n",
      "Calculating cluster Ductal_CRISP3_high-centroacinar_like\n",
      "\n",
      "Calculating cluster Cancer_clone_A\n",
      "\n",
      "Calculating cluster Ductal_MHC_Class_II\n",
      "\n",
      "Calculating cluster Cancer_clone_B\n",
      "\n",
      "Calculating cluster mDCs_A\n",
      "\n",
      "Calculating cluster Ductal_APOL1_high-hypoxic\n",
      "\n",
      "Calculating cluster Tuft_cells\n",
      "\n",
      "Calculating cluster mDCs_B\n",
      "\n",
      "Calculating cluster pDCs\n",
      "\n",
      "Calculating cluster Endocrine_cells\n",
      "\n",
      "Calculating cluster Endothelial_cells\n",
      "\n",
      "Calculating cluster Macrophages_A\n",
      "\n",
      "Calculating cluster Mast_cells\n",
      "\n",
      "Calculating cluster Macrophages_B\n",
      "\n",
      "Calculating cluster T_cells_&_NK_cells\n",
      "\n",
      "Calculating cluster Monocytes\n",
      "\n",
      "Calculating cluster RBCs\n",
      "\n",
      "Calculating cluster Fibroblasts\n",
      "\n"
     ]
    }
   ],
   "source": [
    "scrna_path = 'Spatialsmooth/Example/sc_rna.h5seurat'\n",
    "spatial_path = 'Spatialsmooth/Example/st_data.h5seurat'\n",
    "celltype_final = 'celltype'\n",
    "python_path = \"D:/Application/anaconda3/envs/python_3.10\"\n",
    "output_path = 'output/test_output/'\n",
    "\n",
    "# 运行所有方法\n",
    "run_deconvolution_methods(scrna_path, spatial_path,celltype_final, python_path, output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d6d77f9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
